[NEW] nst_sale_like: PROJECT-2 Step 1 – skeleton

nataliastekolnikovainnowise opened the pull request:
### What’s inside
- Module skeleton for PROJECT-2:
  - __manifest__.py, __init__.py
  - models: nst.order, nst.order.line
  - views: menus, actions, list/form (<list> for Odoo 18)
  - security: ir.model.access.csv

### How to test
1) Add repo path to --addons-path.
2) Install module: -u nst_sale_like on clean DB (18.0).
3) Menu: NST Demo → Orders / Order Lines.

nataliastekolnikovainnowise committed:
[NEW] nst_sale_like: PROJECT-2 Step 1 – skeleton

Create autonomous Odoo 18 addon:
- __manifest__.py, __init__.py
- models: nst.order, nst.order.line
- views: menus, actions, list/form (<list> in Odoo 18)
- security: ir.model.access.csv

nataliastekolnikovainnowise committed:
[IMP] nst_sale_like: PROJECT-2 Step 2 – parent-line architecture

Introduce parent-line relation:
- nst.order with order_line_ids
- nst.order.line with order_id
- Inline editable list view (quantity, price_unit, extra_cost)
- Access rights for base.group_user

nataliastekolnikovainnowise commented:
### Update: Step 2 complete
- Implemented parent ↔ line architecture:
  - nst.order with One2many to nst.order.line
  - Inline editable list (Quantity, Unit Price, Extra Cost)
  - Monetary fields with currency_id
  - Access rights for base.group_user
Tested in Odoo 18: Orders can be created and lines are saved correctly.

nataliastekolnikovainnowise committed:
[FEAT] nst_sale_like: PROJECT-2 Step 3 – compute base_subtotal

Add computed stored field on order lines:
- base_subtotal = quantity * price_unit + extra_cost
- Depends on quantity, price_unit, extra_cost
- Shown as readonly in list and form views

nataliastekolnikovainnowise committed:
[IMP] nst_sale_like: PROJECT-2 Step 4 – summary totals on order

- Added field amount_base on nst.order to aggregate base_subtotal of lines
- Stored compute with proper dependency
- Updated form view: new 'Summary' tab with Base Amount (readonly)
- Updated list view: added Base Amount column

nataliastekolnikovainnowise commented:
### Step 3 complete
- Added compute field `base_subtotal` on order lines:
  - Formula: `quantity * price_unit + extra_cost`
  - Stored field (`store=True`) for performance and correct grouping
  - Currency via related field from order
  -  Tested in Odoo 18: editing lines updates `Base Subtotal`.

nataliastekolnikovainnowise commented:
### Step 4 complete
- Added summary totals on order (parent):
  - `amount_base = sum(line.base_subtotal)`
  - Monetary field with `currency_id`
- Updated views:
  - Added column `Base Amount` in list view
  - Added field `amount_base` in form view under **Summary** tab
  - Tested in Odoo 18: totals recalculate correctly when lines are added/edited.

nataliastekolnikovainnowise committed:
[IMP] nst_sale_like: PROJECT-2 Step 5 – add draft/confirmed states with buttons
- Added field 'state' (draft/confirmed) on nst.order
- Implemented action_confirm() and action_set_draft() methods
- Updated form view: added buttons Confirm and Set to Draft
- Buttons are state-dependent (Confirm visible in draft, Set to Draft visible in confirmed)

nataliastekolnikovainnowise committed:
[REF] nst_sale_like: PROJECT-2 Step 6 – unify quotes style
- Replaced single quotes with double quotes in manifest and  models
- Ensured consistency with coding guidelines

nataliastekolnikovainnowise commented:
Step 5–6 complete

Step 5 – add draft/confirmed states with buttons
    - Added field "state" (draft/confirmed) on nst.order
    - Implemented action_confirm() and action_set_draft() methods
    - Updated form view: added buttons Confirm and Set to Draft
    - Buttons are state-dependent (Confirm visible in draft, Set to Draft visible in confirmed)

Step 6 – unify quotes style
    - Replaced single quotes with double quotes in manifest and models
    - Ensured consistency with coding guidelines

Tested in Odoo 18:
    - Buttons correctly change state and visibility
    - Code follows project coding standards (quotes style unified)

aliaksandrzhaldybin commented on __manifest__.py:
```diff
@@ -0,0 +1,11 @@
+{
+    "name": "NST Sale-Like Demo (Step 1)",
+    "version": "1.0",
+    "depends": ["base", "product"],  # product обязателен
```
contact

nataliastekolnikovainnowise replied:
Добавлен author в __manifest__.py

nataliastekolnikovainnowise replied:
+email

aliaksandrzhaldybin commented on models/nst_order.py:
```diff
@@ -0,0 +1,105 @@
+    name = fields.Char(string="Order Reference")
+
+    customer_id = fields.Many2one(
+        "res.partner",
```
все аргументы именованные

nataliastekolnikovainnowise replied:
Исправлено - заменены позиционные аргументы на именованные во всех полях (comodel_name, string и т.д.)

aliaksandrzhaldybin commented on models/nst_order.py:
```diff
@@ -0,0 +1,105 @@
+        string="Total Amount",
+        compute="_compute_amount_total",
+        store=True,
+        readonly=True,
```
нет смысла делать поле ридонли, т.к оно compute(рид онли по умолчанию)

aliaksandrzhaldybin commented on models/nst_order.py:
```diff
@@ -0,0 +1,105 @@
+            order.state = "draft"
+
+    # === Подсчёт общей суммы ===
+    @api.depends("order_line_ids.subtotal")
```
в депендс должны быть добавлены и просто линии
order_line_ids, order_line_ids.subtotal - правильно будет так
Иначе система не отследит удаление

aliaksandrzhaldybin commented on models/nst_sale_like.py:
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
+from odoo import models, fields
```
зачем этот файл при наличии файла выше?

nataliastekolnikovainnowise replied:
Файл nst_sale_like.py удален из репозитория и убран импорт из __init__.py. Весь функционал находится в nst_order.py

aliaksandrzhaldybin reviewed (changes requested):
see comments

nataliastekolnikovainnowise committed:
[IMP] nst_sale_like: PROJECT-2 Step 7 – add Summary block in form view
- Added fields amount_total, amount_avg, amount_max on nst.order
- Implemented _compute_summary() method
- Updated form view: added new Summary tab with readonly fields

nataliastekolnikovainnowise committed:
[FIX] nst_sale_like: PROJECT-2 Step 7 – replace Monetary with Float
- Removed currency_id dependency
- Changed price_unit and subtotal to Float fields
- Fixed module loading error

nataliastekolnikovainnowise committed:
[IMP] nst_sale_like: PROJECT-2 Step 8 - Add discount functionality
- Added discount field to order lines with percentage calculation
- Added discount_percent field to order summary for manual input
- Implemented subtotal recalculation based on discount (always from original value)
- Added state-based readonly logic: draft allows all edits, confirmed allows only discount edits
- Total amount recalculates automatically when summary discount changes

nataliastekolnikovainnowise committed:
[IMP] nst_sale_like: PROJECT-2 Step 9 - Add field validations
- Added validation for quantity field: cannot be negative
- Added validation for price_unit field: cannot be negative
- Added validation for discount field: cannot be negative
- ValidationError messages displayed to users on constraint violations

nataliastekolnikovainnowise commented:
Step 7 - Summary Block Complete

Added comprehensive order summary functionality on nst.order:

Added computed fields amount_total, amount_avg, amount_max with proper dependencies
Implemented _compute_summary() method with @api.depends on order line subtotals
Updated form view: added dedicated Summary tab with readonly calculated fields
Fields use store=True for performance optimization and database queries
Summary automatically updates when order lines change
Tested in Odoo 18: adding/editing lines immediately updates summary calculations

nataliastekolnikovainnowise commented:
Step 8 - Discount Functionality Complete

Implemented comprehensive discount system for order management:

Added discount field to order lines with percentage-based calculation
Added discount_percent field to order summary for manual override
Subtotal recalculation formula: quantity * price_unit * (1 - discount/100)
Always calculates from original values, never compounds discounts
State-based editing permissions: draft state allows full editing, confirmed state restricts to discount field only
Total amount automatically recalculates when summary discount changes
Currency consistency maintained throughout discount calculations
Tested in Odoo 18: discount changes trigger immediate subtotal and total updates

nataliastekolnikovainnowise commented:
Step 9 - Field Validations Complete

Added comprehensive input validation for numeric fields:

quantity field: prevents negative values with user-friendly error message
price_unit field: ensures non-negative pricing with validation constraint
discount field: blocks negative discount percentages
Implemented using @api.constrains decorators for real-time validation
ValidationError exceptions provide clear feedback to users
Tested in Odoo 18: attempting negative values triggers appropriate error dialogs

nataliastekolnikovainnowise committed:
[IMP] nst_sale_like: PROJECT-2 Step 10 - Fix summary fields discount calculation
- Fixed discount_percent field to recalculate all summary fields properly
- Updated _compute_summary dependencies to include discount_percent field
- Total Amount, Average Amount, and Max Amount now recalculate with discount
- Discount percentage applies to all computed summary values consistently
- Maintained existing validation constraints for discount field

nataliastekolnikovainnowise commented:
**Step 10 - Order Summary Discount Calculation Fix**

Fixed critical issue where discount percentage wasn't properly recalculating all order summary fields:

**Enhanced computation dependencies**: Updated `@api.depends` decorator to include all relevant order line fields (`quantity`, `price_unit`, `discount`) alongside `discount_percent` for comprehensive recalculation triggers

**Corrected discount application logic**: Previously only `amount_total` was affected by order-level discount; now all summary fields (`amount_total`, `amount_avg`, `amount_max`) consistently apply the discount percentage

**Improved code structure**: Replaced direct field assignments with calculated variables (`final_total`, `final_avg`, `final_max`) to ensure consistent discount application across all computed values

**Real-time recalculation**: Changes to any order line field or discount percentage now properly trigger recalculation of all summary statistics, maintaining data consistency throughout the order lifecycle

Tested in Odoo 18: modifying order line quantities, prices, or discount percentages now correctly updates all summary fields with proper discount application. This ensures accurate financial calculations and prevents discrepancies between displayed totals and actual computed values.

nataliastekolnikovainnowise committed:
[FIX] Add author information to manifest

- Added author field with contact information as requested by mentor
- Author: Natalia Stekolnikova <natalia.stekolnikova@innovise.com>

nataliastekolnikovainnowise committed:
[FIX] Use named arguments in field definitions

- Replaced positional arguments with named arguments in all field definitions
- Added comodel_name parameter for Many2one and One2many fields
- Added string parameter for all field labels
- Fixed as requested by mentor review

nataliastekolnikovainnowise committed:
[FIX] Remove nst_sale_like import from models/__init__.py

- Removed import for deleted nst_sale_like.py file as requested by mentor
- Keeps only nst_order import for the complete implementation

nataliastekolnikovainnowise committed:
[IMP] nst_sale_like: PROJECT-2 Step 11 - Implement readonly logic for order states

- Draft status: all order line fields editable, can add/delete lines
- Confirmed status: only discount field remains editable
- Added product_name field to replace product_id dependency
- Removed dependency on product module (now depends only on base)
- Implemented via XML readonly attributes with parent.state conditions
- Maintains data integrity while allowing discount adjustments in confirmed orders
